File: server/app/models/audio.py
from pydantic import BaseModel, Field


class AudioMetadata(BaseModel):
    filename: str
    duration: float
    sample_rate: int
    channels: int
    format: str


class RhythmAnalysis(BaseModel):
    bpm: float
    bpm_confidence: float = Field(ge=0, le=1)
    beats: list[float]
    downbeats: list[float]
    time_signature: int = 4
    tempo_stable: bool = True
    tempo_curve: dict[str, list[float]] | None = None


class EnergyBands(BaseModel):
    bass: list[float]
    low_mid: list[float]
    mid: list[float]
    high_mid: list[float]
    treble: list[float]


class SpectralAnalysis(BaseModel):
    times: list[float]
    rms: list[float]
    spectral_centroid: list[float]
    spectral_flux: list[float]
    spectral_rolloff: list[float]
    mfcc: list[list[float]]
    energy_bands: EnergyBands


class TonalAnalysis(BaseModel):
    key: str
    scale: str
    key_confidence: float = Field(ge=0, le=1)
    chromagram: list[list[float]]


class MoodAnalysis(BaseModel):
    valence: float = Field(ge=-1, le=1)
    energy: float = Field(ge=0, le=1)
    danceability: float = Field(ge=0, le=1)
    tags: list[str]


class HarmonicPercussive(BaseModel):
    harmonic_energy: list[float]
    percussive_energy: list[float]


class SectionData(BaseModel):
    boundaries: list[float]
    labels: list[str]
    confidence: list[float]
    similarities: list[list[float]]


class AudioAnalysisResult(BaseModel):
    metadata: AudioMetadata
    rhythm: RhythmAnalysis
    sections: SectionData
    spectral: SpectralAnalysis
    tonal: TonalAnalysis
    mood: MoodAnalysis
    onsets: list[float]
    harmonic_percussive: HarmonicPercussive

File: client/src/types/audio.ts
export interface AudioMetadata {
  filename: string;
  duration: number;
  sampleRate: number;
  channels: number;
  format: string;
}

export interface RhythmAnalysis {
  bpm: number;
  bpmConfidence: number;
  beats: number[];
  downbeats: number[];
  timeSignature: number;
  tempoStable: boolean;
  tempoCurve?: {
    times: number[];
    bpms: number[];
  };
}

export interface SectionData {
  boundaries: number[];
  labels: string[];
  confidence: number[];
  similarities: number[][];
}

export interface EnergyBands {
  bass: number[];
  lowMid: number[];
  mid: number[];
  highMid: number[];
  treble: number[];
}

export interface SpectralAnalysis {
  times: number[];
  rms: number[];
  spectralCentroid: number[];
  spectralFlux: number[];
  spectralRolloff: number[];
  mfcc: number[][];
  energyBands: EnergyBands;
}

export interface TonalAnalysis {
  key: string;
  scale: string;
  keyConfidence: number;
  chromagram: number[][];
}

export interface MoodAnalysis {
  valence: number;
  energy: number;
  danceability: number;
  tags: string[];
}

export interface HarmonicPercussive {
  harmonicEnergy: number[];
  percussiveEnergy: number[];
}

export interface AudioAnalysis {
  metadata: AudioMetadata;
  rhythm: RhythmAnalysis;
  sections: SectionData;
  spectral: SpectralAnalysis;
  tonal: TonalAnalysis;
  mood: MoodAnalysis;
  onsets: number[];
  harmonicPercussive: HarmonicPercussive;
}

export interface AudioFeaturesAtTime {
  rms: number;
  spectralCentroid: number;
  spectralFlux: number;
  spectralRolloff: number;
  spectralFlatness: number;
  energyBands: {
    bass: number;
    lowMid: number;
    mid: number;
    highMid: number;
    treble: number;
  };
  harmonicEnergy: number;
  percussiveEnergy: number;
  onsetStrength: number;
}

export interface BeatState {
  isOnBeat: boolean;
  beatIndex: number;
  timeSinceLastBeat: number;
  timeToNextBeat: number;
  beatIntensity: number;
}
